File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/FavoriteScreenPresenter.java
package com.recursia.popularmovies.presentation.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.recursia.popularmovies.Screens;
import com.recursia.popularmovies.domain.FavoriteScreenInteractor;
import com.recursia.popularmovies.domain.models.Movie;
import com.recursia.popularmovies.presentation.views.contracts.FavoriteScreenContract;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import ru.terrakok.cicerone.Router;

@InjectViewState
public class FavoriteScreenPresenter extends MvpPresenter<FavoriteScreenContract> {
    private final FavoriteScreenInteractor favoriteScreenInteractor;
    private final Router router;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();

    public FavoriteScreenPresenter(FavoriteScreenInteractor favoriteScreenInteractor, Router router) {
        this.favoriteScreenInteractor = favoriteScreenInteractor;
        this.router = router;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().hideNoFavoriteScreen();
        initMovies();
    }

    private void initMovies() {
        Disposable d = favoriteScreenInteractor.getAllFavoriteMovies()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::handleResult, this::handleError);
        compositeDisposable.add(d);
    }

    private void handleResult(List<Movie> movies) {
        if (!movies.isEmpty()) {
            getViewState().setMovies(movies);
            getViewState().hideNoFavoriteScreen();
        } else {
            getViewState().showNoFavoriteScreen();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.dispose();
    }

    private void handleError(Throwable t) {
        getViewState().showErrorMessage(t.getLocalizedMessage());
    }

    public void onItemClicked(Movie movie) {
        router.navigateTo(new Screens.DetailScreen(movie.getId()));
    }

    public void onBackPressed() {
        router.exit();
    }

}

File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/FavoriteScreenPresenter.kt
package com.recursia.popularmovies.presentation.presenters

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.recursia.popularmovies.Screens
import com.recursia.popularmovies.domain.FavoriteScreenInteractor
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.presentation.views.contracts.FavoriteScreenContract
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import ru.terrakok.cicerone.Router

@InjectViewState
class FavoriteScreenPresenter(private val favoriteScreenInteractor: FavoriteScreenInteractor, private val router: Router) : MvpPresenter<FavoriteScreenContract>() {
    private val compositeDisposable = CompositeDisposable()

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.hideNoFavoriteScreen()
        initMovies()
    }

    private fun initMovies() {
        val d = favoriteScreenInteractor.allFavoriteMovies
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ handleResult(it) }, { handleError(it) })
        compositeDisposable.add(d)
    }

    private fun handleResult(movies: List<Movie>) {
        if (movies.isNotEmpty()) {
            viewState.setMovies(movies)
            viewState.hideNoFavoriteScreen()
        } else {
            viewState.showNoFavoriteScreen()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.dispose()
    }

    private fun handleError(t: Throwable) {
        viewState.showErrorMessage(t.localizedMessage)
    }

    fun onItemClicked(movie: Movie) {
        router.navigateTo(Screens.DetailScreen(movie.id))
    }

    fun onBackPressed() {
        router.exit()
    }

}
